pkg/controller/util.go

```diff
@@ -109,8 +110,158 @@ func readDecodedRole(s string) (*spec.PgUser, error) {
-	if *rolesSecret == (spec.NamespacedName{}) {
+var emptyNamespacedName = (spec.NamespacedName{})
+
+// Return information about what secrets we need to use to create
+// infrastructure roles and in which format are they. This is done in
+// compatible way, so that the previous logic is not changed, and handles both
+// configuration in ConfigMap & CRD.
+func (c *Controller) getInfrastructureRoleDefinitions() []*config.InfrastructureRole {
+	var roleDef config.InfrastructureRole
+	rolesDefs := c.opConfig.InfrastructureRoles
+
+	if c.opConfig.InfrastructureRolesSecretName == emptyNamespacedName {
+		// All the other possibilities require secret name to be present, so if
+		// it is not, then nothing else to be done here.
+		return rolesDefs
+	}
+
+	// check if we can extract something from the configmap config option
+	if c.opConfig.InfrastructureRolesDefs != "" {
+		// The configmap option could contain either a role description (in the
+		// form key1: value1, key2: value2), which has to be used together with
+		// an old secret name.
+
+		propertySep := ","
+		valueSep := ":"
+
+		// The field contains the format in which secret is written, let's
+		// convert it to a proper definition
+		properties := strings.Split(c.opConfig.InfrastructureRolesDefs, propertySep)
+
+		roleDef = config.InfrastructureRole{
+			Secret:   c.opConfig.InfrastructureRolesSecretName,
+			Template: false,
+		}
+
+		for _, property := range properties {
+			values := strings.Split(property, valueSep)
+			if len(values) < 2 {
+				continue
+			}
+			name := strings.TrimSpace(values[0])
+			value := strings.TrimSpace(values[1])
+
+			switch name {
+			case "name":
+				roleDef.Name = value
+			case "password":
+				roleDef.Password = value
+			case "role":
+				roleDef.Role = value
+			default:
+				c.logger.Warningf("Role description is not known: %s",
+					c.opConfig.InfrastructureRolesSecretName)
+			}
+		}
+	} else {
+		// At this point we deal with the old format, let's replicate it
+		// via existing definition structure and remember that it's just a
+		// template, the real values are in user1,password1,inrole1 etc.
+		roleDef = config.InfrastructureRole{
+			Secret:   c.opConfig.InfrastructureRolesSecretName,
+			Name:     "user",
+			Password: "password",
+			Role:     "inrole",
+			Template: true,
+		}
+	}
+
+	if roleDef.Name != "" &&
+		roleDef.Password != "" &&
+		roleDef.Role != "" {
+		rolesDefs = append(rolesDefs, &roleDef)
+	}
+
+	return rolesDefs
+}
+
+func (c *Controller) getInfrastructureRoles(
+	rolesSecrets []*config.InfrastructureRole) (
+	map[string]spec.PgUser, []error) {
+
+	var errors []error
+	var noRolesProvided = true
+
+	roles := []spec.PgUser{}
+	uniqRoles := map[string]spec.PgUser{}
+
+	// To be compatible with the legacy implementation we need to return nil if
+	// the provided secret name is empty. The equivalent situation in the
+	// current implementation is an empty rolesSecrets slice or all its items
+	// are empty.
```

Comment:
Maybe I'm overlooking something, but doesn't `getInfrastructureRoleDefinitions` set the secret name when the old format is used? Can it be empty then?

Reply:
Yes, it sets if from the old field, which apparently could be empty (judging from the previous tests implementation).